src/levels/sparsetrianglelevels.jl

```diff
@@ -121,63 +118,30 @@ end
 virtual_level_eltype(lvl::VirtualSparseTriangleLevel) = virtual_level_eltype(lvl.lvl)
 virtual_level_default(lvl::VirtualSparseTriangleLevel) = virtual_level_default(lvl.lvl)
 
-function sparsetrianglesize2(shape, n)
-    size = 0
-    for dim in 1:n
-        levelsize = 1 
-        for i in 1:dim
-            levelsize = call(*, levelsize, call(+, shape, dim - i - 1))
-        end
-        # size += fld(levelsize, factorial(dim))
-        size = call(+, size, call(fld, levelsize, factorial(dim)))
-    end
-    return call(+, size, 1)
-end
-
 function declare_level!(lvl::VirtualSparseTriangleLevel, ctx::LowerJulia, pos, init)
-    # qos = call(>>>, call(*, call(*, pos, lvl.shape), call(+, lvl.shape, lvl.Ti(1))), lvl.Ti(1))
-    qos = sparsetrianglesize2(lvl.shape, lvl.N)
+    qos = virtual_simplex(lvl.N, ctx, lvl.shape)
     lvl.lvl = declare_level!(lvl.lvl, ctx, qos, init)
     return lvl
 end
 
 function trim_level!(lvl::VirtualSparseTriangleLevel, ctx::LowerJulia, pos)
-    qos = ctx.freshen(:qos)
-    size = ctx.freshen(:size)
-    dim = ctx.freshen(:dim)
-    levelsize = ctx.freshen(:levelsize)
-    i = ctx.freshen(:i)
-
-    push!(ctx.preamble, quote
-        # $qos = (($(ctx(lvl.shape)) * ($(ctx(lvl.shape)) + $(lvl.Ti(1)))) >>> 0x01)
-        # $qos = sparsetrianglesize2($(ctx(lvl.shape)), $(ctx(lvl.N)))
-        $size = 0
-        for $dim in 1:$(ctx(lvl.N))
-            $levelsize = 1 
-            for $i in 1:$dim
-                $levelsize *= $(ctx(lvl.shape)) + $dim - $i - 1
-            end
-            $size += fld($levelsize, factorial($dim))
-        end
-        $qos = $size + 1
-    end)
-    lvl.lvl = trim_level!(lvl.lvl, ctx, value(qos))
+    qos = call(*, pos, virtual_simplex(lvl.N, ctx, lvl.shape))
+    lvl.lvl = trim_level!(lvl.lvl, ctx, qos)
     return lvl
 end
 
 function assemble_level!(lvl::VirtualSparseTriangleLevel, ctx, pos_start, pos_stop)
-    # qos_start = call(+, call(*, call(-, pos_start, lvl.Ti(1)), lvl.shape), 1)
-    # qos_stop = call(>>>, call(*, call(*, pos_stop, lvl.shape), call(+, lvl.shape, lvl.Ti(1))), lvl.Ti(1))
-    lvl_size = sparsetrianglesize2(lvl.shape, lvl.N)
-    qos_start = call(+, call(*, call(-, pos_start, lvl.Ti(1)), lvl_size), 1)
-    qos_stop = call(>>>, call(*, call(*, pos_stop, lvl_size), call(+, lvl_size, lvl.Ti(1))), lvl.Ti(1))
```

Comment:
This line with the >>> is just an optimized simplex when n == 2, so this is not ND. Perhaps we should add a rewrite rule for fld(n, 2) => n >>> 1, though